core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -584,7 +589,31 @@ public UpdateNode plan(Update node)
         builder = planAndMappings.getSubPlan();
 
         ImmutableList.Builder<Symbol> updatedColumnValuesBuilder = ImmutableList.builder();
-        orderedColumnValues.forEach(columnValue -> updatedColumnValuesBuilder.add(planAndMappings.get(columnValue)));
+        Assignments.Builder assignmentsBuilder = Assignments.builder();
+        assignmentsBuilder.putIdentities(builder.getRoot().getOutputSymbols());
+        Map<String, ColumnMetadata> columnNameToMetadata = columnsMetadata.stream()
+                .collect(toImmutableMap(ColumnMetadata::getName, identity()));
+
+        for (int i = 0; i < targetColumnNames.size(); i++) {
+            String columName = targetColumnNames.get(i);
+            ColumnMetadata columnMetadata = columnNameToMetadata.get(columName);
+            Symbol output = symbolAllocator.newSymbol(columName, columnMetadata.getType());
+            updatedColumnValuesBuilder.add(output);
+            Expression expression = ((ProjectNode) builder.getRoot()).getAssignments().get(planAndMappings.get(orderedColumnValues.get(i)));
```

Comment:
This shouldn't assume `builder.getRoot()` is a ProjectNode.

Reply:
Why ? The fact that it is a `ProjectNode` is a result of a call to `coerce` function that is made several lines above
```
PlanAndMappings planAndMappings = coerce(builder, orderedColumnValues, analysis, idAllocator, symbolAllocator, typeCoercion);
```
and coerce method always returns `ProjectNode`
```
subPlan = subPlan.withNewRoot(
                new ProjectNode(
                        idAllocator.getNextId(),
                        subPlan.getRoot(),
                        assignments.build()));

        return new PlanAndMappings(subPlan, mappings);
```

Reply:
@martint I've just tried several different approaches and none of them worked for all the tests I run. If you don't like that one with relying on `ProjectNode` could you please suggest something ?

Reply:
You just need to get the output of the coerce() method for the given input expression:

```java
Expression expression = planAndMappings.get(orderedColumnValues.get(i)).toSymbolReference();
```

In fact, getting the expression from the ProjectNode assignments is wrong -- it's basically undoing whatever the `coerce()` method did by using the underlying expressions. If you add a test that requires coercions, you'll probably see a failure or incorrect plans.

Reply:
ohh I didn't know I can do `toSymbolReference` great, thank you !